### PR TITLE
Added Nock integration for mocking QnA Tests

### PR DESCRIPTION
**Pull request by JuanAr**
Tests in `qnaMaker.test.js` were adapted to use Nock and retrieve responses stored in JSON files.

- Each test is coupled with mocked responses in JSON files named after each test.
- For tests with multiple service calls, mocked responses have an incremental numeric suffix to be queued following that order.
- A test was added to validate a response for a question with no correct answers after receiving a response with correct answers.
- All asserts were changed to `assert.strictEqual`, and an error message was added for each assert.
- Inline explicit timeout was removed.
- In case the `hostname` environment variable is not defined, a default one is used by Nock
- A flag variable `mockQnA` was added to toggle between using the service or the mocked responses

### CODE REVIEW

File: libraries/botbuilder-ai/tests/qnaMaker.test.js [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
 const ai = require('../');
+const nock = require('nock');
+const fs = require('fs');
 
 // Save test keys
 const knowledgeBaseId = process.env.QNAKNOWLEDGEBASEID;
 const endpointKey = process.env.QNAENDPOINTKEY;
-const hostname = process.env.QNAHOSTNAME;
+const hostname = process.env.QNAHOSTNAME || 'botbuilder-test-app';
+const mockQnA = true;
```
[the remark this answers is not in the document]
Comment (JuanAr) [reply libraries/botbuilder-ai/tests/qnaMaker.test.js]: We'll include this logic. Moreover, we'll include an override flag to force using the mocks.

File: libraries/botbuilder-ai/tests/qnaMaker.test.js [hunk 2]
```diff
@@ -18,58 +21,97 @@ class TestContext extends TurnContext {
...
-        console.warn('WARNING: skipping QnAMaker test suite because QNAKNOWLEDGEBASEID environment variable is not defined');
-        return;
+        console.warn('WARNING: QnAMaker test suite QNAKNOWLEDGEBASEID environment variable is not defined');
     }
     if (!endpointKey) {
-        console.warn('WARNING: skipping QnAMaker test suite because QNAENDPOINTKEY environment variable is not defined');
-        return;
+        console.warn('WARNING: QnAMaker test suite QNAENDPOINTKEY environment variable is not defined');
     }
     if (!hostname) {
```
[the remark this answers is not in the document]
Comment (JuanAr) [reply libraries/botbuilder-ai/tests/qnaMaker.test.js]: Ok, we'll remove this.

File: libraries/botbuilder-ai/tests/qnaMaker.test.js [hunk 3]
```diff
@@ -18,58 +21,97 @@ class TestContext extends TurnContext {
...
     it('should work free standing', function () {
         const qna = new ai.QnAMaker(endpoint, { top: 1 });
+        let answer = 'BaseCamp: You can use a damp rag to clean around the Power Pack';
 
         return qna.generateAnswer(`how do I clean the stove?`)
             .then(res => {
-                assert(res);
-                assert(res.length == 1);
-                assert(res[0].answer.startsWith("BaseCamp: You can use a damp rag to clean around the Power Pack"));
+                assert.strictEqual(typeof res !== 'undefined', true, 'The response was returned as \'undefined\'.');
```
[the remark this answers is not in the document]
Comment (JuanAr) [reply libraries/botbuilder-ai/tests/qnaMaker.test.js]: We'll use the second option, to keep consistency between all the asserts.

File: libraries/botbuilder-ai/tests/qnaMaker.test.js [hunk 4]
```diff
@@ -18,58 +21,97 @@ class TestContext extends TurnContext {
...
+    });
+
+    it('should return 0 answers for a question with no answer after a succesful call', function () {
+        const qna = new ai.QnAMaker(endpoint, { top: 1 });
+        let answer = 'BaseCamp: You can use a damp rag to clean around the Power Pack';
+
+        return qna.generateAnswer(`how do I clean the stove?`)
+            .then(res => {
+                assert.strictEqual(typeof res !== 'undefined', true, 'The response was returned as \'undefined\'.');
+                assert.strictEqual(res.length, 1, 'Should have receive just one answer on the first call.');
```
[the remark this answers is not in the document]
Comment (JuanAr) [reply libraries/botbuilder-ai/tests/qnaMaker.test.js]: We'll fix this typo.

**Comment by JuanAr** [comment]
@stevengum we applied your feedback. Thanks!

**Comment by stevengum** [comment]
@JuanAr, can you look into why the build is failing, and also include an explicit console log that indicates if the mocks (nocks?) are running.

**Pull request merged**